lib/modem_key_mgmt/modem_key_mgmt.c

```diff
@@ -0,0 +1,236 @@
+		return -EFAULT;
+	}
+
+	if (!cmee_was_active) {
+		if (cmee_set(true)) {
+			return -EIO;
+		}
+	}
+
+	retval = at_cmd_write(cmd, buf, buf_len, state);
+
+	if (!cmee_was_active) {
+		if (cmee_set(false)) {
+			return -EIO;
+		}
+	}
+
+	return retval;
+}
+
+int modem_key_mgmt_write(nrf_sec_tag_t sec_tag,
+			 enum modem_key_mgnt_cred_type_t cred_type,
+			 const u8_t *buffer, u16_t buffer_len)
+{
+	int result;
+	int written;
+	enum at_cmd_state state;
+
+	if ((buffer == NULL) || (buffer_len == 0)) {
+		return -EINVAL;
+	}
+
+	written = snprintf(scratch_buf, sizeof(scratch_buf), "%s,%u,%u,\"",
+			   MODEM_KEY_MGMT_OP_WR, (u32_t)sec_tag,
+			   (u8_t)cred_type);
+	if ((written < 0) || (written >= sizeof(scratch_buf))) {
+		return -ENOBUFS;
+	}
+
+	if ((written + buffer_len + sizeof("\"\r\n")) > sizeof(scratch_buf)) {
+		return -ENOBUFS;
+	}
+
+	memcpy(&scratch_buf[written], buffer, buffer_len);
+	written += buffer_len;
+
+	memcpy(&scratch_buf[written], "\"\r\n", sizeof("\"\r\n"));
+
+	result = write_at_cmd_with_cme_enabled(scratch_buf, NULL, 0, &state);
+	if ((result != 0) && (state == AT_CMD_ERROR_CME)) {
+		return translate_error(result);
+	}
+
+	return result;
+}
+
+int modem_key_mgmt_read(nrf_sec_tag_t sec_tag,
+			enum modem_key_mgnt_cred_type_t cred_type, u8_t *buffer,
+			u16_t *buffer_len)
+{
+	int result;
+	int written;
+	enum at_cmd_state state;
+
+	if ((buffer == NULL) || (buffer_len == NULL) || (*buffer_len == 0)) {
+		return -EINVAL;
+	}
+
+	written = snprintf(scratch_buf, sizeof(scratch_buf), "%s,%u,%u\r\n",
+			   MODEM_KEY_MGMT_OP_RD, (u32_t)sec_tag,
+			   (u8_t)cred_type);
+	if ((written < 0) || (written >= sizeof(scratch_buf))) {
+		return -ENOBUFS;
+	}
+
+	result = write_at_cmd_with_cme_enabled(scratch_buf, scratch_buf,
+					       sizeof(scratch_buf), &state);
+	if ((result != 0) && (state == AT_CMD_ERROR_CME)) {
+		return translate_error(result);
+	}
+
+	if (result == 0) {
+		size_t len;
+		struct at_param_list cmng_list;
+
+		at_params_list_init(&cmng_list, AT_CMNG_PARAMS_COUNT);
+		at_parser_params_from_str(scratch_buf, NULL, &cmng_list);
+		at_params_size_get(&cmng_list, AT_CMNG_CONTENT_INDEX, &len);
+		if (len < *buffer_len) {
+			at_params_list_free(&cmng_list);
+			return -ENOBUFS;
+		}
+		at_params_string_get(&cmng_list, AT_CMNG_CONTENT_INDEX, buffer,
+				     &len);
+		*buffer_len = len;
+		at_params_list_free(&cmng_list);
+	}
+
+	return result;
+}
+
+int modem_key_mgmt_delete(nrf_sec_tag_t sec_tag,
+			  enum modem_key_mgnt_cred_type_t cred_type)
+{
+	int result;
+	int written;
+	enum at_cmd_state state;
+
+	written = snprintf(scratch_buf, sizeof(scratch_buf), "%s,%u,%u\r\n",
+			   MODEM_KEY_MGMT_OP_RM, (u32_t)sec_tag,
+			   (u8_t)cred_type);
+	if ((written < 0) || (written >= sizeof(scratch_buf))) {
+		return -ENOBUFS;
+	}
+
+	result = write_at_cmd_with_cme_enabled(scratch_buf, NULL, 0, &state);
+	if ((result != 0) && (state == AT_CMD_ERROR_CME)) {
+		return translate_error(result);
```

Comment:
My comment on line 63 would make it possible to `return translate_error` without additional logic.